cmd/gateway/hdfs/gateway-hdfs.go

```diff
@@ -0,0 +1,706 @@
+	}
+
+	return uploadID, nil
+}
+
+func (n *hdfsObjects) ListMultipartUploads(ctx context.Context, bucket string, prefix string, keyMarker string, uploadIDMarker string, delimiter string, maxUploads int) (lmi minio.ListMultipartsInfo, err error) {
+	_, err = n.clnt.Stat(minio.PathJoin(hdfsSeparator, bucket))
+	if err != nil {
+		return lmi, hdfsToObjectErr(ctx, err, bucket)
+	}
+
+	// It's decided not to support List Multipart Uploads, hence returning empty result.
+	return lmi, nil
+}
+
+func (n *hdfsObjects) checkUploadIDExists(ctx context.Context, bucket, object, uploadID string) (err error) {
+	_, err = n.clnt.Stat(minio.PathJoin(hdfsSeparator, minioMetaTmpBucket, uploadID))
+	if err != nil {
+		return hdfsToObjectErr(ctx, err, bucket, object, uploadID)
+	}
+	return nil
+}
+
+func (n *hdfsObjects) ListObjectParts(ctx context.Context, bucket, object, uploadID string, partNumberMarker int, maxParts int, opts minio.ObjectOptions) (result minio.ListPartsInfo, err error) {
+	_, err = n.clnt.Stat(minio.PathJoin(hdfsSeparator, bucket))
+	if err != nil {
+		return result, hdfsToObjectErr(ctx, err, bucket)
+	}
+
+	if err = n.checkUploadIDExists(ctx, bucket, object, uploadID); err != nil {
+		return result, err
+	}
+
+	// It's decided not to support List parts, hence returning empty result.
+	return result, nil
+}
+
+func (n *hdfsObjects) CopyObjectPart(ctx context.Context, srcBucket, srcObject, dstBucket, dstObject, uploadID string, partID int,
+	startOffset int64, length int64, srcInfo minio.ObjectInfo, srcOpts, dstOpts minio.ObjectOptions) (minio.PartInfo, error) {
+	return n.PutObjectPart(ctx, dstBucket, dstObject, uploadID, partID, srcInfo.PutObjReader, dstOpts)
+}
+
+func (n *hdfsObjects) PutObjectPart(ctx context.Context, bucket, object, uploadID string, partID int, r *minio.PutObjReader, opts minio.ObjectOptions) (info minio.PartInfo, err error) {
+	_, err = n.clnt.Stat(minio.PathJoin(hdfsSeparator, bucket))
+	if err != nil {
+		return info, hdfsToObjectErr(ctx, err, bucket)
+	}
+
+	var w *hdfs.FileWriter
+	w, err = n.clnt.Append(minio.PathJoin(hdfsSeparator, minioMetaTmpBucket, uploadID))
+	if err != nil {
+		return info, hdfsToObjectErr(ctx, err, bucket, object, uploadID)
+	}
+
+	_, err = io.Copy(w, r.Reader)
+	if err != nil {
+		w.Close()
+		return info, hdfsToObjectErr(ctx, err, bucket, object, uploadID)
+	}
+	w.Close()
+
+	info.PartNumber = partID
+	info.ETag = r.MD5CurrentHexString()
+	info.LastModified = minio.UTCNow()
+	info.Size = r.Reader.Size()
+
+	return info, nil
+}
+
+func (n *hdfsObjects) CompleteMultipartUpload(ctx context.Context, bucket, object, uploadID string, parts []minio.CompletePart, opts minio.ObjectOptions) (objInfo minio.ObjectInfo, err error) {
+	_, err = n.clnt.Stat(minio.PathJoin(hdfsSeparator, bucket))
+	if err != nil {
+		return objInfo, hdfsToObjectErr(ctx, err, bucket)
+	}
+
+	if err = n.checkUploadIDExists(ctx, bucket, object, uploadID); err != nil {
+		return objInfo, err
+	}
```

Comment:
Able to complete multipart with incorrect part ETag - are we not going to verify part ETag's here?

Reply:
yes we cannot verify any of that right now due to lack of xattr support